### PR TITLE
Fixes #1219 : Added IsEqualToOrdering Classifier Rule 

### LINKED ISSUE #1219: Add classifiers isEqualToOrdering
Add new classifier rule `isEqualToOrdering`

**Target PR date**: 22 May 2020
**Target completion date**: 24 May 2020

### PR DESCRIPTION
**Pull request by aggarwalpulkit596**
## Explanation

Fixes #1219 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: @BenHenning is that okay if I do it like this?

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: Please add EOF newline.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Please add spacing between the ``,``s (see other code in the codebase for consistency). In general, I suggest running the linter on all the new code your add to make sure it's following our code style. Ditto elsewhere.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Nit: here & elsewhere, no spaces between ``(`` and their nearby elements unless the left of it is a keyword. See: https://developer.android.com/kotlin/style-guide#horizontal.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Nit: here & elsewhere, prefer importing ``assertThat`` directly (since it's meant to read like a sentence, and 'Truth' distracts that a bit).

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: I find it a bit hard to understand what the actual structure will look like for these answers. Can we be more explicit when setting them up? E.g. could we pass in a list of lists and assemble it that way?

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: Here & elsewhere: please make sure all of the spacing you're using follows our style guide and matches other code in the codebase.

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: Can you add a test to verify that different lengths of the outer list also fail as expected? As well as another for the ordering of the outer list (versus just an inner list).

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: Is this doing a set & count comparison? It might be a bit simpler to approach it that way. For example:

```kotlin
private fun areListOfSetsOfHtmlStringsEqual(answer: ListOfSetsOfHtmlStrings, input: ListOfSetsOfHtmlStrings): Boolean {
  if (answer.setOfHtmlStringsCount != input.setOfHtmlStringsCount) {
    return false
  }
  val answerStringSets = answer.setOfHtmlStringsList
  val inputStringSets = answer.setOfHtmlStringsList
  return (answerStringSets zip inputStringSets).map { (first, second) ->
    areSetsOfHtmlStringsEqual(first, second)
  }.reduce(Boolean::and)
}

private fun areSetsOfHtmlStringsEqual(first: StringList, second: StringList): Boolean {
  return HashSet<>(first.htmlList) == HashSet<>(second.htmlList)
}
```

Also, I was just matching the existing behavior. Is it the case that we do care about the ordering of the outer list, but not the ordering of the inner lists?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @aggarwalpulkit596. Had a few comments.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (vinitamurthi) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: FYI You can run ktlint on a specific file this way: ` ktlint -F <path_to_your_file>/fileName.kt`

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (vinitamurthi) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Add a blank line at the EOF

**Review by vinitamurthi — APPROVED** [review]
I will defer to Ben for the classification logic. Left some nits

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: @BenHenning this will only this ordering of the first list but not the inner list so we have to check the ordering of inner list as well correct?

Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:50]: It’s probably worth adding a brief comment here explaining the logic. It’s complicated enough to follow that a basic example comment could help clarify things.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Prefer not having wildcard imports per the style guide: https://developer.android.com/kotlin/style-guide#import_statements.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: Maybe move the lists to their own constants and reuse them everywhere rather than reconstructing them. You can also name the test constants to provide a bit more context where they're used, e.g.:

```kotlin
private val ITEM_SET_1_AB = listOf("item a", "item b")
private val ITEM_SET_2_ITEM_2 = listOf("item 2")
private val ITEM_SET_3_ITEM_3 = listOf("item 3")
private val ITEM_SET_4_INVALID_AB = listOf("item invalid a", "item invalid b")
...
private val LIST_OF_SETS_123 = createListOfSetsOfHtmlStrings(ITEM_SET_1_AB, ITEM_SET_2...)
```

This will also help with the ordering test below, because you could have a ``LIST_OF_SETS_213`` to differentiate the ordering of the sets.

**Review by BenHenning — APPROVED** [review]
Thanks @aggarwalpulkit596! General PR LGTM, but I added a few comments to try to improve clarity/context in some places.

**Comment by BenHenning** [comment]
Also: please add the issue being fixed to the PR title & indicate that this is blocked on your other PR being submitted first.

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:50]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt]: done

**Comment by BenHenning** [comment]
@aggarwalpulkit596 see #2 https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#tips-for-getting-your-pr-merged-after-submission. This PR should not have been merged--there are still open comment threads. Please refrain from merging PRs in the future until all reviewers approve the PR, and all comment threads have been resolved.

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:37]: This is not a KDoc, so it should just be a standard block comment (``/*`` not ``/**``), or a series of single line comments (which is what I suggested before I think).

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:38]: Missing space between ``*`` and ``For``.

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt:33]: This doesn't quite solve my comment from earlier. When you're reading the test, it's still not clear why VALUE_1 versus VALUE_2 contain elements in different orders (in fact, it seems like they're entirely different values which requires readers to look at the constants to reason about whether the test is correct). To indicate ordering, we would need to include the actual element names & their orders in the constant name. Please update the constants to include their names, or inline the constants so that the context is directly in the test body.

**Review by BenHenning — COMMENTED** [review]
@aggarwalpulkit596 please resolve all open comments in this PR in a new one since this one was merged.

**Comment by aggarwalpulkit596** [comment]
@BenHenning I thought you approved this PR that's why i merged this but will keep this in mind for future PR

File: domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt
Comment (aggarwalpulkit596) [reply domain/src/test/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProviderTest.kt:33]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:38]: done

File: domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt
Comment (aggarwalpulkit596) [reply domain/src/main/java/org/oppia/domain/classify/rules/dragAndDropSortInput/DragDropSortInputIsEqualToOrderingClassifierProvider.kt:37]: done

**Pull request merged**
